heartbeat/monitors/stdfields/unnest.go

```diff
@@ -0,0 +1,53 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package stdfields
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// OptionalStream represents a config that has a stream set, which in practice
+// means agent/fleet. In this case we primarily use the first stream for the
+// config, but we do pull the Id from the root, and merge the root data stream
+// in as well
+type OptionalStream struct {
+	Id         string           `config:"id"`
+	DataStream *common.Config   `config:"data_stream"`
+	Streams    []*common.Config `config:"streams"`
+}
+
+// UnnestStream detects configs that come from fleet and transforms the config into something compatible
+// with heartbeat, by mixing some fields (id, data_stream) with those from the first stream. It assumes
+// that there is exactly one stream associated with the input.
+func UnnestStream(config *common.Config) (res *common.Config, err error) {
```

Comment:
I wonder if we can rather have agent (or the spec file) to produce a complete/correct configuration (for all Beats). The merging code here makes it look like the responsibility of understanding streams from Fleet configurations is shared between agent+heartbeat.

Reply:
I'm not sure if I know the answer to that since I'm unfamiliar with how the parsing for the other beats works. For heartbeat we decided to keep the agent logic minimal and put more in heartbeat. That decision has worked out well since debugging issues in heartbeat is much simpler without fleet involved. I know that effectively `heartbeat.monitors` is equivalent to `inputs` in agent, so I can copy / paste fleet configs straight in for manual testing to confirm.